do_py/data_object/dynamic_restrictions.py

```diff
@@ -0,0 +1,167 @@
+"""
+Dynamic restrictions for a DataObject.
+:date_created: 2020-07-10
+:author: Gian Brazzini
+"""
+from do_py import DataObject, R
+from do_py.abc import ABCRestrictionMeta
+from do_py.data_object.restriction import ManagedRestrictions, _ListValueRestriction
+from do_py.utils.properties import cached_property
+
+
+class DynamicRestrictions(ManagedRestrictions):
+    """
+    A dict that contains the dynamic restrictions where the key is the independent
+    key's value and the value is the dependent key's restriction.
+    """
+    _restriction = R()
+
+    def manage(self):
+        """
+        Validate that the list is not empty and that the format is correct.
+        """
+        assert self.data, 'Missing dynamic restrictions, dict cannot be empty.'
+        for value in self.data.values():
+            assert DataObject in value.mro(), 'Dynamic restrictions depend each value being a DataObject.'
+
+
+class DynamicClassGenerator(DataObject):
+    """
+    Generate a class that is inheritable by a DO that requires dynamic restrictions.
+    :restriction independent_key: The key that the dynamic restriction depends on.
+    :restriction dependent_key: The key where the restriction is dynamic.
+    :restriction dynamic_restrictions: See `DynamicRestrictions` ManagedList.
+    """
+    _restrictions = {
+        'independent_key': R.STR,
+        'dependent_key': R.STR,
+        'dynamic_restrictions': DynamicRestrictions()
+        }
+
+    @property
+    def update_fn_name(self):
+        """
+        :rtype: str
+        """
+        return '_update_%s_restriction' % self.dependent_key
+
+    @cached_property
+    def dynamic_class(self):
+        """
+        Create a class dynamically which inherits from DataObject, but also contains the required attributes to
+        create dynamic restrictions for the dependent key.
+        :rtype: type(DataObject)
```

Comment:
`:rtype: ABCRestrictionMeta`

Reply:
I agree. I put it like this because pycharm was giving me a headache when using ABCRestrictionMeta